src/compiler/new.ts

```diff
@@ -184,64 +220,75 @@ export function compileNewExpression(state: CompilerState, node: ts.NewExpressio
 		let result = `{`;
 		if (args.length === 1) {
 			const arg = args[0];
-			if (
-				ts.TypeGuards.isNumericLiteral(arg) &&
-				arg.getText().match(/^\d+$/) &&
-				arg.getLiteralValue() <= ARRAY_NIL_LIMIT
-			) {
-				const literalValue = arg.getLiteralValue();
-				if (literalValue !== 0) {
-					result += ", nil".repeat(literalValue).substring(1) + " ";
-				}
-			} else {
-				throw new CompilerError(
-					"Invalid argument #1 passed into ArrayConstructor. Expected a simple integer fewer or equal to " +
-						ARRAY_NIL_LIMIT +
-						".",
-					node,
-					CompilerErrorType.BadBuiltinConstructorCall,
-				);
-			}
+			result = "table.create(" + compileExpression(state, arg) + ")";
+			return appendDeclarationIfMissing(
+				state,
+				skipNodesUpwards(node.getParent()),
+				result
+			);
+		} else if (args.length === 2) {
+			const [length, value] = args;
+			result =
+				"table.create(" +
+				compileExpression(state, length) +
```

Comment:
You shouldn't manually have logic here for putting a comma between 2 compiled expressions, since we have to do other checks on `length` and `value` here. You need to use compileCallArguments, otherwise, the preceding expressions won't be handled properly.

Reply:
Alright, wasn't aware of that. Will fix.